include/circt/Dialect/ESI/ESITypes.h

```diff
@@ -0,0 +1,39 @@
+//===- ESITypes.h - types for the ESI dialect -------------------*- C++ -*-===//
+//
+// Types for ESI are mostly in tablegen. This file should contain C++ types used
+// in MLIR type parameters.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_ESI_ESITYPES_HPP
+#define CIRCT_DIALECT_ESI_ESITYPES_HPP
+
+#include "mlir/IR/Dialect.h"
+#include "mlir/IR/Operation.h"
+
+#include "ESIDialect.h"
+
+namespace circt {
+namespace esi {
+
+// A field in a struct or union
+struct FieldInfo {
+  llvm::StringRef name;
+  mlir::Type type;
+
+public:
+  FieldInfo(llvm::StringRef name, mlir::Type type) : name(name), type(type) {}
+
+  FieldInfo allocateInto(::mlir::TypeStorageAllocator &alloc) const {
+    llvm::StringRef nameCopy = alloc.copyInto(name);
+    return FieldInfo(nameCopy, type);
+  }
```

Comment:
I don't see this used anywhere, and (as I think you've discovered) using StringRef's this way is a bit of a dangerous pattern.  I think that MLIR has a string uniquing system in `MLIRContext`, I'd recommend using that.